rust/src/mime/mime.rs

```diff
@@ -0,0 +1,581 @@
+    let mut errored = false;
+    let mut input = i;
+    while !input.is_empty() {
+        match take_until::<_, &[u8], nom7::error::Error<&[u8]>>("\r\n")(input) {
+            Ok((input2, line)) => {
+                match mime_parse_header_line(line) {
+                    Ok((value, name)) => {
+                        if slice_equals_lowercase(name, "content-disposition".as_bytes()) {
+                            let mut sections_values = Vec::new();
+                            if let Some(filename) = mime_find_header_token(
+                                value,
+                                "filename".as_bytes(),
+                                &mut sections_values,
+                            ) {
+                                if !filename.is_empty() {
+                                    ctx.filename = Vec::with_capacity(filename.len());
+                                    fileopen = true;
+                                    for c in filename {
+                                        // unescape
+                                        if *c != b'\\' {
+                                            ctx.filename.push(*c);
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                        if value.is_empty() {
+                            errored = true;
+                        }
+                    }
+                    _ => {
+                        if !line.is_empty() {
+                            errored = true;
+                        }
+                    }
+                }
+                let (input3, _) = tag("\r\n")(input2)?;
+                input = input3;
+                if line.is_empty() || (line.len() == 1 && line[0] == b'\r') {
+                    return Ok((input, (MimeParserState::HeaderEnd, fileopen, errored)));
+                }
+            }
+            _ => {
+                // guard against too long header lines
+                if input.len() > MIME_HEADER_MAX_LINE {
+                    return Ok((
+                        input,
+                        (MimeParserState::BoundaryWaitingForEol, fileopen, errored),
+                    ));
+                }
+                if input.len() < i.len() {
+                    return Ok((input, (MimeParserState::Header, fileopen, errored)));
+                } // else only an incomplete line, ask for more
+                return Err(Err::Error(make_error(input, ErrorKind::Eof)));
+            }
+        }
+    }
+    return Ok((input, (MimeParserState::Header, fileopen, errored)));
+}
+
+type NomTakeError<'a> = Err<nom7::error::Error<&'a [u8]>>;
+
+fn mime_consume_chunk<'a>(boundary: &[u8], input: &'a [u8]) -> IResult<&'a [u8], bool> {
+    let r: Result<(&[u8], &[u8]), NomTakeError> = take_until("\r\n")(input);
+    match r {
+        Ok((input, line)) => {
+            let (input2, _) = tag("\r\n")(input)?;
```

Comment:
Naming? I hit this naming issue with nom as well often.

Reply:
renaming to `next_line`. Should I add more comments ?